File: runners/nodejs/src/interfaces/processor.ts
import { RunnerError } from "../error";
import { Arguments } from "../runtime/arguments";

export abstract class Processor {
  /* Retrieve a processor definition by its resource name. */
  protected args: Arguments;

  /* Parse the incoming arguments. */
  constructor(args: Arguments) {
    this.args = args;
  }

  /* The actual implementation of the processor must be overridden here. */
  public async exec(): Promise<void> {
    throw RunnerError.missingImplementation();
  }
}

File: orchestrator/rdfc-processor/src/main/kotlin/Processor.kt
package technology.idlab.rdfc.processor

import technology.idlab.rdfc.core.log.Log

/**
 * Processors are the building blocks of RDF-Connect. They are responsible for executing the actual
 * logic inside a pipeline. This class models a processor targeting the JVMRunner
 */
abstract class Processor(
    /**
     * The arguments of a processor are stored in a map and can be accessed by name. Type safety is
     * assured through the use of Kotlin reflection.
     */
    protected val arguments: Arguments,
) {
  @JvmField val log = Log.shared

  /**
   * The `exec` function will be called when the pipeline executes. All inter-processor
   * communication must happen here, and not in the constructor.
   */
  abstract suspend fun exec()
}
